op-node/sources/engine_client.go

```diff
@@ -0,0 +1,110 @@
+package sources
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/sources/caching"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type EngineClientConfig struct {
+	L2ClientConfig
+}
+
+func EngineClientDefaultConfig(config *rollup.Config) *EngineClientConfig {
+	return &EngineClientConfig{
+		// engine is trusted, no need to recompute responses etc.
+		L2ClientConfig: *L2ClientDefaultConfig(config, true),
+	}
+}
+
+// EngineClient extends L2Client with engine API bindings.
+type EngineClient struct {
+	*L2Client
+}
+
+func NewEngineClient(client client.RPC, log log.Logger, metrics caching.Metrics, config *EngineClientConfig) (*EngineClient, error) {
+	l2Client, err := NewL2Client(client, log, metrics, &config.L2ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EngineClient{
+		L2Client: l2Client,
+	}, nil
+}
+
+// ForkchoiceUpdate updates the forkchoice on the execution client. If attributes is not nil, the engine client will also begin building a block
+// based on attributes after the new head block and return the payload ID.
+// May return an error in ForkChoiceResult, but the error is marshalled into the error return
```

Comment:
The statement about `ForkChoiceResult` error could be a bit clearer. Where is the error being marshalled?

Reply:
Good catch, it's legacy from when we still tried to present the non-success status values with a regular error. But we don't anymore. I'll update the description to clarify the exact error usage.

Reply:
Will also update the `NewPayload` doc :+1: